File: 1021-remove-outermost-parentheses/1021-remove-outermost-parentheses.java
class Solution {
    public String removeOuterParentheses(String s) {
        StringBuilder result = new StringBuilder();
        int open = 0, close = 0, start = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') open++;
            else if (s.charAt(i) == ')') close++;
            if (open == close) {
                result.append(s.substring(start + 1, i));
                start = i + 1;
            }
        }
        return result.toString();
    }
}

File: 1021-remove-outermost-parentheses/1021-remove-outermost-parentheses.cpp
class Solution {
public:
    string removeOuterParentheses(string s) {
        string res;
        
        for (int i = 0, n = s.size(), open = 0; i < n; i++) {
            if (s[i] == '(') {
                if (open > 0) {
                    res = res + '(';
                }
                open++;
            } else {
                if (open > 1) {
                    res = res + ')';
                }
                open--;
            }
        }
        
        return res;
    }
};

/*

while adding '(' we are checking if there is already an open bracket in stock, if yes then this is not the outermost one so we can add it to res

same for ')' if there is 1 opening bracket in stock then it will be for the outermost pair, so we need more than 1 open bracket for this close bracket matching
*/